src/ynEIGEN/LSDWrapper.sol

```diff
@@ -0,0 +1,78 @@
+import {IERC4626} from "lib/openzeppelin-contracts/contracts/interfaces/IERC4626.sol";
+import {IERC20, SafeERC20} from "lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IwstETH} from "src/external/lido/IwstETH.sol";
+import {IWrapper} from "src/interfaces/IWrapper.sol";
+
+contract LSDWrapper is IWrapper, Initializable {
+
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable wstETH;
+    IERC20 public immutable woETH;
+    IERC20 public immutable oETH;
+    IERC20 public immutable stETH;
+
+    // ============================================================================================
+    // Constructor
+    // ============================================================================================
+
+    constructor(address _wstETH, address _woETH, address _oETH, address _stETH) {
+        if (_wstETH == address(0) || _woETH == address(0) || _oETH == address(0) || _stETH == address(0)) {
+            revert ZeroAddress();
+        }
+
+        wstETH = IERC20(_wstETH);
+        woETH = IERC20(_woETH);
+        oETH = IERC20(_oETH);
+        stETH = IERC20(_stETH);
+    }
+
+    function initialize() external initializer {
+        stETH.forceApprove(address(wstETH), type(uint256).max);
+        oETH.forceApprove(address(woETH), type(uint256).max);
+    }
+
+    // ============================================================================================
+    // External functions
+    // ============================================================================================
+
+    /// @inheritdoc IWrapper
+    function wrap(uint256 _amount, IERC20 _token) external returns (uint256, IERC20) {
+        if (_token == stETH) {
+            stETH.safeTransferFrom(msg.sender, address(this), _amount);
+            _amount = IwstETH(address(wstETH)).wrap(_amount);
+            wstETH.safeTransfer(msg.sender, _amount);
+            return (_amount, wstETH);
+        } else if (_token == oETH) {
+            oETH.safeTransferFrom(msg.sender, address(this), _amount);
+            return (IERC4626(address(woETH)).deposit(_amount, msg.sender), woETH);
+        } else {
+            return (_amount, _token);
+        }
+    }
+
+    /// @inheritdoc IWrapper
+    function unwrap(uint256 _amount, IERC20 _token) external returns (uint256, IERC20) {
+        if (_token == wstETH) {
+            wstETH.safeTransferFrom(msg.sender, address(this), _amount);
+            _amount = IwstETH(address(wstETH)).unwrap(_amount);
+            stETH.safeTransfer(msg.sender, _amount);
+            return (_amount, stETH);
+        } else if (_token == woETH) {
+            return (IERC4626(address(woETH)).redeem(_amount, msg.sender, msg.sender), oETH);
+        } else {
+            return (_amount, _token);
+        }
+    }
+
+    // ============================================================================================
```

Comment:
move `toUserAssetAmount` logic in here

Reply:
https://github.com/yieldnest/yieldnest-protocol/pull/164/commits/2de43b785a001986a5cc10abb76e0fca2974f70d